### PR TITLE
Add apop2john.py to extract POP3 APOP traffic. Implements #4882

### PR DESCRIPTION
**Pull request by marksilinio**
https://www.root-me.org/en/Challenges/Network/POP-APOP can be used for testing

#4882

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
Please, create a PR and provide [sample files](https://github.com/openwall/john-samples) so we can test.

**Comment by marksilinio** [comment]
> Please, create a PR and provide [sample files](https://github.com/openwall/john-samples) so we can test.

see https://github.com/openwall/john-samples/pull/8

**Review by ghost — APPROVED** [review]

File: run/apop2john.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python
+
+# This software is Copyright (c) 2021 Mark Silinio <mark.silinio-at-gmail.com>,
+# and it is hereby released to the general public under the following terms:
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted.
+#
+# Utility to bruteforce APOP hash
```
Comment (solardiz) [inline run/apop2john.py]: I think we need to change this comment and the similar one in `radius2john.py` to say something more reasonable like, "Extract and format POP3 APOP challenge/responses for password cracking with JtR" here and "Extract and format suitable RADIUS packets for shared secret cracking with JtR" there.

**Review by solardiz — APPROVED** [review]
What versions of Python has this been tested with? Is this known to be compatible or incompatible with Python 2, or not known either way?

I ask because we currently use explicit `python3` on the shebang line for scripts that are known to require Python 3. This one says simply `python`, so (given our current convention) claims to work with both Python 2 and 3. Ditto for `radius2john.py`, so if any fix is needed for both scripts then it can be in a separate PR.

**Comment by exploide** [comment]
I'm wondering: Would it make sense to wrap this into a function and incorporate it in [pcap2john.py](https://github.com/openwall/john/blob/bleeding-jumbo/run/pcap2john.py) instead of having a new standalone `apop2john.py` script?

Most other hash extraction logic that operates on pcap files is bundled there. Only few other standalone pcap parsers exist.

Having it in `pcap2john.py` has the advantage that one does not need to know beforehand what types of hashes the pcap file contains. It just extracts everything it knows about.

**Comment by solardiz** [comment]
> Would it make sense to wrap this into a function and incorporate it in [pcap2john.py](https://github.com/openwall/john/blob/bleeding-jumbo/run/pcap2john.py) instead of having a new standalone `apop2john.py` script?

I'd support this, but unfortunately I find `pcap2john.py` somewhat problematic: it has extra hard dependency on `dpkt` and no way to choose to extract traffic of just one protocol. (And for a multi protocol extractor, it's also somewhat inefficient in that it goes over the input multiple times, once per protocol. Since that's how it works anyway, it feels silly that there's no way to force it to just one protocol when desired.)

So for now this PR should stay as-is - adding a new script. We already have `radius2john.py` anyway.

BTW, @marksilinio this is a significant enough contributions that I think you should add a `doc/NEWS` entry for it (then we'll revise the entry if we merge the script into `pcap2john.py` before the next release).

**Review by solardiz — CHANGES_REQUESTED** [review]
The reference to 4883 in the commit message is wrong, should be 4882 or nothing. Please correct.

**Comment by solardiz** [comment]
I'd use "Add apop2john.py to extract POP3 APOP traffic" for the commit message.

**Comment by exploide** [comment]
> I'd support this, but unfortunately I find pcap2john.py somewhat problematic: it has extra hard dependency on dpkt and no way to choose to extract traffic of just one protocol. (And for a multi protocol extractor, it's also somewhat inefficient in that it goes over the input multiple times, once per protocol. Since that's how it works anyway, it feels silly that there's no way to force it to just one protocol when desired.)

Fully agree. Having two network packet parsing libaries (dpkt and scapy) ist not very clean. Most of the extractors use dpkt, however I consider scapy being more powerful. It could also simplify portions of the code.

I also thought about the inefficiency, you are right. It could be improved in multiple ways. I can for example imagine that each extractor function is refactored in a class that holds the logic and also offers a filter function. In the main function, each packet would be checked against each filter function, that is supposed to be fast (only check protocol, port number or the like) and is passed to the extractor only if applicable. No need to re-read the whole pcap file in each and every extractor.

But this requires a lot of work. I think I don't find the time right now. But if anyone is interested, I'm willing to brain-storm further.

But this goes off-topic here.

**Comment by marksilinio** [comment]
> I think we need to change this comment and the similar one in `radius2john.py` to say something more reasonable like, "Extract and format POP3 APOP challenge/responses for password cracking with JtR" here and "Extract and format suitable RADIUS packets for shared secret cracking with JtR" there.

Updated

> What versions of Python has this been tested with? Is this known to be compatible or incompatible with Python 2, or not known either way?

Updated - now it both python2/python3 compatible:
```
x@xxx:~/marksilinio/github/john$ python2.7 run/apop2john.py ../../tmp/apop-ch.pcapng 
billcl:$dynamic_1017$5aadb7d8a2639088a7036fe6d5cd5727$HEX$3c313733332e312e39663430333632352e4262574767704b7a5a4b4b4c386f7067576e307775413d3d407670732d61653961626337323e
x@xxx:~/marksilinio/github/john$ python3 run/apop2john.py ../../tmp/apop-ch.pcapng 
billcl:$dynamic_1017$5aadb7d8a2639088a7036fe6d5cd5727$HEX$3c313733332e312e39663430333632352e4262574767704b7a5a4b4b4c386f7067576e307775413d3d407670732d61653961626337323e
```

> The reference to 4883 in the commit message is wrong, should be 4882 or nothing. Please correct.

fixed

> I'd use "Add apop2john.py to extract POP3 APOP traffic" for the commit message.

fixed

**Review by solardiz — APPROVED** [review]
Looks good to me. Thank you!

**Pull request merged**